fpdf/fpdf.py

```diff
@@ -358,6 +361,34 @@ def __init__(
         # final buffer holding the PDF document in-memory - defined only after calling output():
         self.buffer = None
 
+    def set_encryption(
+        self,
+        owner_password,
+        user_password=None,
+        encryption_method=None,
+        permissions=AccessPermission.all(),
+        encrypt_metadata=False,
+    ):
+
+        """
+        A file ID is needed for the security handler to generate the password hashes
+        Override file_id method so it guarantees the same id will be generated at the trailer
+        """
+        fid = self._default_file_id(bytearray([0xFF]))
```

Comment:
This hack bothers me:
* it does not take into account custom `file_id()` methods that can be defined by users,
and can even shadow them (could you please add a unit test combining a custom `file_id()` & `set_encryption()`?)
* assigning a function to an attribute is not the same as defining a method, and we ran into trouble when mixing those two approaches in the past (notably when `FPDFRecorder` is used)

I think we should figure a way to avoid this hack,
but it seems almost impossible to me to use an ID based on the PDF content,
as the encryption key depens on it and must be built before the document is serialized (and the buffer content known).

Hence the best option may be to ask/require users to specify a `file_id()` method themselves
when they use `set_encryption()`. What do you think?

Reply:
I'll postpone password generation until output() is invoked, so custom file_id() can be used and taken in account